File: ValidBST/bst_validator.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool isBST(struct TreeNode*, int, int);

bool isBST(struct TreeNode* root, int min, int max) {
    if (root == NULL)
        return true;
    if (root && root->left)
        if (root->val == INT_MIN && (root->left->val == INT_MIN) || (root->left->val == INT_MAX))
            return false;
    if (root && root->right)
        if (root->val == INT_MAX && (root->right->val == INT_MAX) || (root->right->val == INT_MIN))
            return false;
    if (root->val <= max && root->val >= min) {
        return isBST(root->left, min, root->val-1) && isBST(root->right, root->val+1, max);
    }
    else
        return false;
}

bool isValidBST(struct TreeNode* root) {
    return isBST(root, INT_MIN, INT_MAX);
}

File: ValidBST/bst_validator.java
public class Solution {
  public boolean isValidBST (TreeNode root) {
    return isValidBST (root, Long.MIN_VALUE, Long.MAX_VALUE);
  }

  public boolean isValidBST (TreeNode root, long minVal, long maxVal) {
    if (root == null) return true;
    // condition that makes it false
    if (root.val >= maxVal || root.val <= minVal) return false;
    /*
    *  at each level, there's a new minVal and maxVal
    * in case of left traversal, maxVal will be that of root (all small nodes to the left)
    * in case of right traversal, minVal will be that of root (all big nodes to the right)
    */
    return isValidBST(root.left, minVal, root.val) && isValidBST(root.right, root.val, maxVal);
  }
}
